### PR TITLE
examples/log: add a option -i to ignore case

### PR DESCRIPTION
**Pull request by pranitbauva1997**
A commonly used option -i to ignore case when used with --grep
to find patterns in commits.

Signed-off-by: Pranit Bauva pranit.bauva@gmail.com

### CODE REVIEW

**Comment by carlosmn** [comment]
The same question as in #3189 and #3398. How widespread is `strcasestr()`?

**Comment by pranitbauva1997** [comment]
@carlosmn My bad! strcasestr() is not widespread. It depends on systems and is not present in many systems and has some dependencies. Since this is a "no dependence" software, I have written my own kind of strcasestr() namely str_ignorecase_str() whose efficiency is not as good as its counter part.

File: examples/log.c [hunk 1]
```diff
@@ -477,3 +490,35 @@ static int parse_options(
...
+
+char *tolowerstr(const char *str) {
+	char *final;
+	int len = 0;
+	int i = 0;
+	while(str[i++] != '\0')
+		;
+	len = i;
+
+	final = (char *) malloc(sizeof(char) * len);
```
Comment (ethomson) [inline examples/log.c]: A stylistic tip - please don't cast `malloc`.  http://stackoverflow.com/questions/605845/do-i-cast-the-result-of-malloc

Also, you do not need `sizeof(char)` here.  That's `1`.

**Comment by ethomson** [comment]
Although I'm not super concerned about the efficiency of our `strcasestr`, it seems like we should be able to get by without `malloc`ing and copying the results?

File: examples/log.c (hunk 1 above)
Comment (pranitbauva1997) [inline examples/log.c]: Oh! I wasn't aware about the malloc thing and sizeof(char). I was under the impression that the size of char is dependent on the compiler. But I just found that its defined in the standard and its always 1. Thanks!

**Comment by pranitbauva1997** [comment]
@ethomson I tried your suggestion. It just gave a warning when I compiled on my system. I thought its okay. But this gives problem with others. Should I just revert back to the previous version?

**Comment by ethomson** [comment]
Sorry if I was unclear - I wasn't suggesting that you change the case of the string in-place; indeed, changing the case of the commit message in-place is not appropriate.

I meant that you should be able to implement a `strcasestr` that works in place, without changing the case of the strings...  a very inefficient implementation might look something like:

```
for (; *haystack; haystack++) {
    char *n, *h;
    bool match = true;

    for (n = needle, h = haystack; *n && *h; n++, h++) {
        if (*n != *h)
            break;
    }

    if (*n != *h)
        continue;

    return haystack;
}
```

(Note this is not tested, I'm just coding in a web browser.)

**Comment by ethomson** [comment]
Hi @pranitbauva1997 - are you interested in trying a `strcasestr` that will not modify the contents of the buffer passed to it?

**Comment by pranitbauva1997** [comment]
@ethomson I am sorry I almost forgot about this PR. I was busy with git. I will look into this during the weekend. Thanks for reminding!

**Comment by ethomson** [comment]
This is quite outdated, and I'm closing this to tidy up the outstanding pull requests.  Thanks for working on this!

**Pull request closed without merging**